man/mysql_config.1
'\" t
.\"     Title: \fBmysql_config\fR
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 02/04/2010
.\"    Manual: MySQL Database System
.\"    Source: MySQL 5.1
.\"  Language: English
.\"
.TH "\FBMYSQL_CONFIG\FR" "1" "02/04/2010" "MySQL 5\&.1" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.\" mysql_config
.SH "NAME"
mysql_config \- get compile options for compiling clients
.SH "SYNOPSIS"
.HP \w'\fBmysql_config\ \fR\fB\fIoptions\fR\fR\ 'u
\fBmysql_config \fR\fB\fIoptions\fR\fR
.SH "DESCRIPTION"
.PP
\fBmysql_config\fR
provides you with useful information for compiling your MySQL client and connecting it to MySQL\&.
.PP
\fBmysql_config\fR
supports the following options\&.
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql_config: cflags option
.\" cflags option: mysql_config
\fB\-\-cflags\fR
.sp
Compiler flags to find include files and critical compiler flags and defines used when compiling the
libmysqlclient
library\&. The options returned are tied to the specific compiler that was used when the library was created and might clash with the settings for your own compiler\&. Use
\fB\-\-include\fR
for more portable options that contain only include paths\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql_config: include option
.\" include option: mysql_config
\fB\-\-include\fR
.sp
Compiler options to find MySQL include files\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql_config: libmysqld-libs option
.\" libmysqld-libs option: mysql_config
.\" mysql_config: embedded option
.\" embedded option: mysql_config
\fB\-\-libmysqld\-libs\fR,
\fB\-\-embedded\fR
.sp
Libraries and options required to link with the MySQL embedded server\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql_config: libs option
.\" libs option: mysql_config
\fB\-\-libs\fR
.sp
Libraries and options required to link with the MySQL client library\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql_config: libs_r option
.\" libs_r option: mysql_config
\fB\-\-libs_r\fR
.sp
Libraries and options required to link with the thread\-safe MySQL client library\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql_config: plugindir option
.\" plugindir option: mysql_config
\fB\-\-plugindir\fR
.sp
The default plugin directory path name, defined when configuring MySQL\&. This option was added in MySQL 5\&.1\&.24\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql_config: port option
.\" port option: mysql_config
\fB\-\-port\fR
.sp
The default TCP/IP port number, defined when configuring MySQL\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql_config: socket option
.\" socket option: mysql_config
\fB\-\-socket\fR
.sp
The default Unix socket file, defined when configuring MySQL\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql_config: version option
.\" version option: mysql_config
\fB\-\-version\fR
.sp
Version number for the MySQL distribution\&.
.RE
.PP
If you invoke
\fBmysql_config\fR
with no options, it displays a list of all options that it supports, and their values:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBmysql_config\fR
Usage: /usr/local/mysql/bin/mysql_config [options]
Options:
  \-\-cflags         [\-I/usr/local/mysql/include/mysql \-mcpu=pentiumpro]
  \-\-include        [\-I/usr/local/mysql/include/mysql]
  \-\-libs           [\-L/usr/local/mysql/lib/mysql \-lmysqlclient \-lz
                    \-lcrypt \-lnsl \-lm \-L/usr/lib \-lssl \-lcrypto]
  \-\-libs_r         [\-L/usr/local/mysql/lib/mysql \-lmysqlclient_r
                    \-lpthread \-lz \-lcrypt \-lnsl \-lm \-lpthread]
  \-\-socket         [/tmp/mysql\&.sock]
  \-\-port           [3306]
  \-\-version        [4\&.0\&.16]
  \-\-libmysqld\-libs [\-L/usr/local/mysql/lib/mysql \-lmysqld \-lpthread \-lz
                    \-lcrypt \-lnsl \-lm \-lpthread \-lrt]
.fi
.if n \{\
.RE
.\}
.PP
You can use
\fBmysql_config\fR
within a command line to include the value that it displays for a particular option\&. For example, to compile a MySQL client program, use
\fBmysql_config\fR
as follows:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBCFG=/usr/local/mysql/bin/mysql_config\fR
shell> \fBsh \-c "gcc \-o progname `$CFG \-\-include` progname\&.c `$CFG \-\-libs`"\fR
.fi
.if n \{\
.RE
.\}
.PP
When you use
\fBmysql_config\fR
this way, be sure to invoke it within backtick (\(lq`\(rq) characters\&. That tells the shell to execute it and substitute its output into the surrounding command\&.
.SH "COPYRIGHT"
.br
.PP
Copyright 2007-2008 MySQL AB, 2008-2010 Sun Microsystems, Inc.
.PP
This documentation is free software; you can redistribute it and/or modify it only under the terms of the GNU General Public License as published by the Free Software Foundation; version 2 of the License.
.PP
This documentation is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along with the program; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA or see http://www.gnu.org/licenses/.
.sp
.SH "SEE ALSO"
For more information, please refer to the MySQL Reference Manual,
which may already be installed locally and which is also available
online at http://dev.mysql.com/doc/.
.SH AUTHOR
Sun Microsystems, Inc. (http://www.mysql.com/).


ChangeLog
------------------------------------------------------------
revno: 3193
tags: clone-5.1.41-build
committer: Georgi Kodinov <joro@sun.com>
branch nick: merge-5.1-bugteam
timestamp: Wed 2009-11-04 11:21:52 +0200
message:
  merge
------------------------------------------------------------
    revno: 1810.3961.16
    committer: Georgi Kodinov <joro@sun.com>
    branch nick: merge-5.0-bugteam
    timestamp: Wed 2009-11-04 11:13:55 +0200
message:
      version change
------------------------------------------------------------
revno: 3192
committer: Georgi Kodinov <joro@sun.com>
branch nick: merge-5.1-bugteam
timestamp: Wed 2009-11-04 11:18:34 +0200
message:
  merge
    ------------------------------------------------------------
    revno: 1810.3961.15
    committer: Georgi Kodinov <joro@sun.com>
    branch nick: merge-5.0-bugteam
    timestamp: Wed 2009-11-04 11:13:22 +0200
    message:
      Disabled the rpl_killed_ddl test in 5.0 because of bug #45520
------------------------------------------------------------
revno: 3191
committer: Davi Arnaut <Davi.Arnaut@Sun.COM>
branch nick: mysql-5.1-bugteam
timestamp: Tue 2009-11-03 16:18:44 -0200
message:
  Automerge.
    ------------------------------------------------------------
    revno: 1810.3961.14
    committer: Davi Arnaut <Davi.Arnaut@Sun.COM>
    branch nick: mysql-5.0-bugteam
    timestamp: Tue 2009-11-03 16:17:06 -0200
    message:
      Fix compiler warnings -- tag unused parameters.
    ------------------------------------------------------------
revno: 3190
committer: Konstantin Osipov <kostja@sun.com>
branch nick: 5.1-41756
timestamp: Tue 2009-11-03 20:45:52 +0300
    message:
  A fix and a test case for 
  Bug#41756 "Strange error messages about locks from InnoDB".
        
  In JT_EQ_REF (join_read_key()) access method, 
  don't try to unlock rows in the handler, unless certain that 
  a) they were locked
  b) they are not used.
  
  Unlocking of rows is done by the logic of the nested join loop,
  and is unaware of the possible caching that the access method may
  have. This could lead to double unlocking, when a row
  was unlocked first after reading into the cache, and then 
  when taken from cache, as well as to unlocking of rows which
  were actually used (but taken from cache).
        
  Delegate part of the unlocking logic to the access method,
  and in JT_EQ_REF count how many times a record was actually 
  used in the join. Unlock it only if it's usage count is 0.
  
  Implemented review comments.
        ------------------------------------------------------------
    revno: 1810.3961.13
    committer: Konstantin Osipov <kostja@sun.com>
    branch nick: 5.0-41756
    timestamp: Tue 2009-11-03 19:58:54 +0300
        message:
      A fix and a test case for
      Bug#41756 "Strange error messages about locks from InnoDB".
          
      In JT_EQ_REF (join_read_key()) access method,
      don't try to unlock rows in the handler, unless certain that
      a) they were locked
      b) they are not used.
          
      Unlocking of rows is done by the logic of the nested join loop,
      and is unaware of the possible caching that the access method may
      have. This could lead to double unlocking, when a row
      was unlocked first after reading into the cache, and then
      when taken from cache, as well as to unlocking of rows which
      were actually used (but taken from cache).
      
      Delegate part of the unlocking logic to the access method,
      and in JT_EQ_REF count how many times a record was actually
      used in the join. Unlock it only if it's usage count is 0.
      
      Implemented review comments.
    ------------------------------------------------------------
revno: 3189
committer: Magnus Bl?udd <magnus.blaudd@sun.com>
branch nick: 5.1-bugteam
timestamp: Tue 2009-11-03 18:07:19 +0100
    message:
  Merge bug#47867 to 5.1-bugteam
        ------------------------------------------------------------
    revno: 3146.2.1
    committer: Magnus Bl?udd <magnus.blaudd@sun.com>
    branch nick: 5.1-bug47867
    timestamp: Tue 2009-10-06 18:10:34 +0200
        message:
      Bug#47867 compiler warning _WIN32_WINNT macro redefinition
       - Remove the conflicting macro definitions since we define a higher value
         directly in the cmake files.
    ------------------------------------------------------------
revno: 3188
committer: Kristofer Pettersson <kristofer.pettersson@sun.com>
branch nick: mysql-5.1-bugteam
timestamp: Tue 2009-11-03 17:23:05 +0100
    message:
  automerge
        ------------------------------------------------------------
    revno: 3148.8.8
    committer: Sergey Vojtovich <svoj@sun.com>
    branch nick: mysql-5.1-bugteam-push
    timestamp: Tue 2009-11-03 18:44:39 +0400
        message:
      Merge of innodb-zip-ss6129 snapshot.
        ------------------------------------------------------------
        revno: 3148.13.37
        committer: Sergey Vojtovich <svoj@sun.com>
        branch nick: mysql-5.1-bugteam
        timestamp: Tue 2009-11-03 18:41:04 +0400
        message:
          Clean-up after applying innodb-zip-ss6129 snapshot:
          - re-enabled main.innodb_bug44369;
          - re-enabled main.innodb_bug47777;
          - re-enabled innodb.innodb_information_schema.
    ------------------------------------------------------------
        revno: 3148.13.36
        committer: Sergey Vojtovich <svoj@sun.com>
        branch nick: mysql-5.1-bugteam
        timestamp: Tue 2009-11-03 14:34:38 +0400
    message:
          Applying InnoDB plugin snashot 
            
          Detailed revision comments:
            
          r6126 | vasil | 2009-10-30 10:36:07 +0200 (Fri, 30 Oct 2009) | 45 lines
          branches/zip: Merge r6112:6125 from branches/5.1:
            
          (skipping r6122 and r6123, Jimmy says these are already present and need
          not be merged):
            

scripts/mysql_config.sh
#!/bin/sh
# Copyright (c) 2000-2008 MySQL AB, 2008 Sun Microsystems, Inc.
# Use is subject to license terms.
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 2 of the License.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA

# This script reports various configuration settings that may be needed
# when using the MySQL client library.

which ()
{
  IFS="${IFS=   }"; save_ifs="$IFS"; IFS=':'
  for file
  do
    for dir in $PATH
    do
      if test -f $dir/$file
      then
        echo "$dir/$file"
        continue 2
      fi
    done
    echo "which: no $file in ($PATH)"
    exit 1
  done
  IFS="$save_ifs"
}

#
# If we can find the given directory relatively to where mysql_config is
# we should use this instead of the incompiled one.
# This is to ensure that this script also works with the binary MySQL
# version

fix_path ()
{
  var=$1
  shift
  for filename
  do
    path=$basedir/$filename
    if [ -d "$path" ] ;
    then
      eval "$var"=$path
      return
    fi
  done
}

get_full_path ()
{
  file=$1

  # if the file is a symlink, try to resolve it
  if [ -h $file ];
  then
    file=`ls -l $file | awk '{ print $NF }'`
  fi

  case $file in
    /*) echo "$file";;
    */*) tmp=`pwd`/$file; echo $tmp | sed -e 's;/\./;/;' ;;
    *) which $file ;;
  esac
}

me=`get_full_path $0`

basedir=`echo $me | sed -e 's;/bin/mysql_config;;'`

ldata='@localstatedir@'
execdir='@libexecdir@'
bindir='@bindir@'

# If installed, search for the compiled in directory first (might be "lib64")
pkglibdir='@pkglibdir@'
pkglibdir_rel=`echo $pkglibdir | sed -e "s;^$basedir/;;"`

plugindir='@pkgplugindir@'

pkgincludedir='@pkgincludedir@'
fix_path pkgincludedir include/mysql include

version='@VERSION@'
socket='@MYSQL_UNIX_ADDR@'
ldflags='@LDFLAGS@'

if [ @MYSQL_TCP_PORT_DEFAULT@ -eq 0 ]; then
  port=0
else
  port=@MYSQL_TCP_PORT@
fi

# Create options 
# We intentionally add a space to the beginning and end of lib strings, simplifies replace later
libs=" $ldflags -L$pkglibdir -lmysqlclient @ZLIB_DEPS@ @NON_THREADED_LIBS@"
libs="$libs @openssl_libs@ @STATIC_NSS_FLAGS@ "
libs_r=" $ldflags -L$pkglibdir -lmysqlclient_r @ZLIB_DEPS@ @LIBS@ @openssl_libs@ "
embedded_libs=" $ldflags -L$pkglibdir -lmysqld @LIBDL@ @ZLIB_DEPS@ @LIBS@ @WRAPLIBS@ @innodb_system_libs@ @openssl_libs@ "

if [ -r "$pkglibdir/libmygcc.a" ]; then
  # When linking against the static library with a different version of GCC
  # from what was used to compile the library, some symbols may not be defined
  # automatically.  We package the libmygcc.a from the build host, to provide
  # definitions for those.  Bugs 4921, 19561, 19817, 21158, etc.
  libs="$libs -lmygcc "
  libs_r="$libs_r -lmygcc "
  embedded_libs="$embedded_libs -lmygcc "
fi

cflags="-I$pkgincludedir @CFLAGS@ " #note: end space!
include="-I$pkgincludedir"

# Remove some options that a client doesn't have to care about
# FIXME until we have a --cxxflags, we need to remove -Xa
#       and -xstrconst to make --cflags usable for Sun Forte C++
# FIXME until we have a --cxxflags, we need to remove -AC99
#       to make --cflags usable for HP C++ (aCC)
for remove in DDBUG_OFF DSAFEMALLOC USAFEMALLOC DSAFE_MUTEX \
              DPEDANTIC_SAFEMALLOC DUNIV_MUST_NOT_INLINE DFORCE_INIT_OF_VARS \
              DEXTRA_DEBUG DHAVE_purify O 'O[0-9]' 'xO[0-9]' 'W[-A-Za-z]*' \
              'mtune=[-A-Za-z0-9]*' 'mcpu=[-A-Za-z0-9]*' 'march=[-A-Za-z0-9]*' \
              Xa xstrconst "xc99=none" AC99 \
              unroll2 ip mp restrict
do
  # The first option we might strip will always have a space before it because
  # we set -I$pkgincludedir as the first option
  cflags=`echo "$cflags"|sed -e "s/ -$remove  */ /g"` 
done
cflags=`echo "$cflags"|sed -e 's/ *\$//'` 

# Same for --libs(_r)
for remove in lmtmalloc static-libcxa i-static static-intel
do
  # We know the strings starts with a space
  libs=`echo "$libs"|sed -e "s/ -$remove  */ /g"` 
  libs_r=`echo "$libs_r"|sed -e "s/ -$remove  */ /g"` 
  embedded_libs=`echo "$embedded_libs"|sed -e "s/ -$remove  */ /g"` 
done

# Strip trailing and ending space if any, and '+' (FIXME why?)
libs=`echo "$libs" | sed -e 's;  \+; ;g' | sed -e 's;^ *;;' | sed -e 's; *\$;;'`
libs_r=`echo "$libs_r" | sed -e 's;  \+; ;g' | sed -e 's;^ *;;' | sed -e 's; *\$;;'`
embedded_libs=`echo "$embedded_libs" | sed -e 's;  \+; ;g' | sed -e 's;^ *;;' | sed -e 's; *\$;;'`

usage () {
        cat <<EOF
Usage: $0 [OPTIONS]
Options:
        --cflags         [$cflags]
        --include        [$include]
        --libs           [$libs]
        --libs_r         [$libs_r]
        --plugindir      [$plugindir]
        --socket         [$socket]
        --port           [$port]
        --version        [$version]
        --libmysqld-libs [$embedded_libs]
EOF
        exit 1
}

if test $# -le 0; then usage; fi

while test $# -gt 0; do
        case $1 in
        --cflags)  echo "$cflags" ;;
        --include) echo "$include" ;;
        --libs)    echo "$libs" ;;
        --libs_r)  echo "$libs_r" ;;
        --plugindir) echo "$plugindir" ;;
        --socket)  echo "$socket" ;;
        --port)    echo "$port" ;;
        --version) echo "$version" ;;
        --embedded-libs | --embedded | --libmysqld-libs) echo "$embedded_libs" ;;
        *)         usage ;;
        esac

        shift
done

#echo "ldata: '"$ldata"'"
#echo "execdir: '"$execdir"'"
#echo "bindir: '"$bindir"'"
#echo "pkglibdir: '"$pkglibdir"'"
#echo "pkgincludedir: '"$pkgincludedir"'"
#echo "version: '"$version"'"
#echo "socket: '"$socket"'"
#echo "port: '"$port"'"
#echo "ldflags: '"$ldflags"'"
#echo "client_libs: '"$client_libs"'"

exit 0
